### ISSUE TITLE
Display of ACC alternate headwords

### ISSUE THREAD

**Issue by drdhaval2785** [sanskrit-lexicon/Cologne#143]
After the incorporation of acc_hwextra.txt into the algorithm flow, the display also needs to be fixed.
Jim asked me to test the display.

There are a couple of issues in advanced search (webtc2).
1. The entries in advanced search are duplicated. There are two 'nAmAparADa'. Both of them have data regarding ID 44337 and ID 47948.1. Ideally there should be only one entry seen with two L-IDs shown. Duplication needs to be removed.
2. Some tags line `<alt>` and `<hwtype>` are still lingering here.

![capture](https://cloud.githubusercontent.com/assets/5755132/26282679/a527edf4-3e33-11e7-84fe-c0f68b7cbe83.png)

**Comment by funderburkjim** [comment]
@drdhaval2785    Have you updated disp.php?  I don't see a modified disp.php in repository.

Pretty Please, SLOWLY ... one step at a time, else it gets confusing with two cooks stirring the same pot.

**Comment by funderburkjim** [comment]
re nAmAparADa  duplication.

By alternate headword generation procedure (i.e. acc_hwextra.txt) we have L=47948.1, nAmAparADa   as
an alternate (of sevAparADa).

However, unexpectedly, the printed text already has an entry for this alternate  (at L=44337).

Thus, we *might* view the 47948.1 entry as superfluous.  If so, then the L=47948.1 line needs to be removed from acc_hwextra.txt.   Then update_sync.sh to recompute everything.

If we think 47948.1 should be removed, then we might want to search for other similar cases in acc_hwextra, and remove them also.

**Comment by drdhaval2785** [comment]
@drdhaval2785 <https://github.com/drdhaval2785> Have you updated disp.php?
I don't see a modified disp.php in repository.


No I thought you did it. Some comment asked me to test display etc. Maybe I
misunderstood.

Pretty Please, SLOWLY ... one step at a time, else it gets confusing with
two cooks stirring the same pot.

I know. I try to rein my emotions, but sometimes they break the barriers.
ACC will be a bit messy, because we are developing new protocols. Once the
protocol is settled, I will try to cook one dictionary by myself, with you
in advisory role. That will upgrade my technical knowledge and confidence
in handling cologne dictionaries.

**Comment by funderburkjim** [comment]
> No I thought you did it

That's fine.   I'll do it for acc.

**Comment by drdhaval2785** [comment]
>re nAmAparADa duplication.

Jim, the issue is a bit deeper than that. There are two entries. One alternate and one regular. If both are seen, I dont have a problem. Problem is that they are both seen twice. Check yourself. I think it is a bug. Basic display is just fine. Only advance has issue.

**Comment by funderburkjim** [comment]
@drdhaval2785   OK.  I'll look more closely tomorrow after altering disp.php.

**Comment by gasyoun** [comment]
> Ideally there should be only one entry seen with two L-IDs shown. 

Agree.

> However, unexpectedly, the printed text already has an entry for this alternate (at L=44337).

That's where it gets interesting.

>  One alternate and one regular. If both are seen, I dont have a problem. Problem is that they are both seen twice.

I've seen it before. It's a more general bug of advanced search, @funderburkjim

**Comment by funderburkjim** [comment]
>  a more general bug of advanced search

Yes.  

A change made to display_outopt4 function in webtc2/query.php (on dev server).

Check out the example above on [dev Adv. Search](http://107.170.79.53/cologne/acc/web/webtc2/index.php).

Now there is only one nAmAparADa.

#### discussion of problem and solution.

There are two records in acc.xml with key1=nAmAparADa;  this particular case is due to the addition of an alternate headword.  But the fact that the second headword is such an alternate is not the cause of the problem.  A similar situation could occur in any dictionary where there two or more records with the same key1.

So when the filtering  (prefix = nAmAparaD)  goes through the records, it correctly finds two matching records with the same key1.

However, when the list of records in the left hand pane is generated, the records with duplicate key1 need to be removed.  That's what the change to query.php accomplishes.

This change should be expected to work in the query.php for other dictionaries, once their query.php files are similarly modified.

**Comment by funderburkjim** [comment]
@drdhaval2785  @gasyoun 

Are there in outstanding issues that need to be addressed with dev version of ACC?
If not, I'll begin the installation of dev-changes on Cologne server, and update the documentation (mainly acc-meta2.txt).

**Comment by drdhaval2785** [comment]
For prefix search, the problem is now resolved. But the problem still
persists for exact search of nAmAparADa.

**Comment by drdhaval2785** [comment]
suffix search mode also gives duplicate results.

There are many corrections which are ongoing in ACC, but they can be added
later on. It is OK to add the current ACC developments to Cologne servers
after fixing the exact search and suffix search items.

One suggestion.

Current wording
(नामापराध is an alternate spelling of सेवापराध.)

Suggested wording
(नामापराध is an alternate of सेवापराध.)

**Comment by funderburkjim** [comment]
A subtlety of the php array_search function had to be taken into account.
This has been done, and (fingers crossed) the exclusion of duplicates now functions properly.

#### discussion
To test whether a given key was already present in an array of keys:
```
$asval = array_search($key,$keys);
OLD WRONG
if ($asval != false) { // then key is in keys}
  When key is the 0th element of keys,  $asval is this index, 0.  0 is treated as false in this comparison.
   In other words, the logic thinks that key is NOT in keys, which is wrong.
NEW RIGHT
if ($asval !== false) { // then key is in keys}
  Using !==  (strict comparison) gives what we want 0  is not false, i.e., key IS in keys.
```
Very tricky.

**Comment by funderburkjim** [comment]
> 'is alternate of'

Changed make_xml.py to this wording. update_sync.sh rerun.

**Comment by funderburkjim** [comment]
#### display of ^x as superscript

Digitization codes superscripts   as xxx^y   (y is superscript in print).
Only instances are when y = a,b,c,d,0,2 and th.

disp.php changed  to use html `<sup>y</sup>` .  This looks prettier.

Sample word: agastyasaMhitA

**Comment by gasyoun** [comment]
> (नामापराध is an alternate of सेवापराध.)

Gets too mystic. Alternate = synonym?

**Comment by funderburkjim** [comment]
All these changes now transferred from dev server to cologne server.
